src/components/Map/index.tsx

```diff
@@ -0,0 +1,190 @@
+		map.setZoom(DEFAULT_ZOOM_LEVEL);
+	} else {
+		const bounds = getMapBounds(markers);
+		map.fitBounds(bounds);
+	}
+
+	// These options depend on the global "google" object, so we set it once the API has loaded.
+	map.setOptions({
+		zoomControlOptions: {
+			position: google.maps.ControlPosition.TOP_RIGHT,
+			style: google.maps.ZoomControlStyle.SMALL,
+		},
+	});
+
+	// If we run `setZoom` right after `fitBounds` the map won't refresh. With this we first wait for the map to be idle (from fitBounds), and then set the zoom level.
+	const listener = google.maps.event.addListenerOnce(map, 'idle', () => {
+		// Don't allow to zoom closer than the defailt detail zoom level on initial load.
+		if (map.getZoom() > DETAIL_ZOOM_LEVEL) {
+			map.setZoom(DETAIL_ZOOM_LEVEL);
+		}
+
+		google.maps.event.removeListener(listener);
+	});
+};
+
+const getFieldFromMap = <T extends any, U extends any>(
+	entry: T,
+	fieldName: keyof MapProps<T>['dataMap'],
+	dataMap: MapProps<T>['dataMap'],
+) => {
+	const field = dataMap[fieldName].find((x) => !!entry[x]);
+	return field ? (entry[field] as U) : undefined;
+};
+
+const Map = <T extends any>({
+	className,
+	data = [],
+	dataMap,
+	getIcon,
+	onItemClick,
+	apiKey,
+	mapClick,
+}: MapProps<T>) => {
+	const markers = React.useMemo(
+		() =>
+			data
+				.map((entry) => {
+					const marker: Partial<UIMarker> = {
+						lng: getFieldFromMap(entry, 'lng', dataMap),
+						lat: getFieldFromMap(entry, 'lat', dataMap),
+						id: getFieldFromMap(entry, 'id', dataMap),
+						title: getFieldFromMap(entry, 'title', dataMap),
+						icon: getIcon ? getIcon(entry) : undefined,
+						click: onItemClick ? () => onItemClick(entry) : undefined,
+					};
+
+					if (!marker.lng || !marker.lat || !marker.id) {
+						return null;
+					}
+
+					return marker;
+				})
+				.filter((x) => !!x) as UIMarker[],
+		[data, dataMap, getIcon, onItemClick],
+	);
+
+	return (
+		<Box height="100%" className={className}>
+			{apiKey && (
+				<LoadScript
+					googleMapsApiKey={apiKey}
+					version="3.35"
+					language="en"
+					preventGoogleFontsLoading
+				>
+					<GoogleMap
+						mapContainerStyle={{
+							height: '100%',
+							minHeight: '300px',
+							opacity: 1,
+						}}
+						options={defaultMapOptions}
+						onLoad={(map) => apiIsLoaded(map, markers)}
+						onClick={mapClick}
+					>
+						{markers.map((marker) => (
+							<Marker
+								key={marker.id}
+								position={{
+									lat: isNaN(marker.lat) ? 0 : marker.lat,
+									lng: isNaN(marker.lng) ? 0 : marker.lng,
+								}}
+								clickable={markers.length > 1}
+								onClick={marker.click}
+								title={marker.title}
+								icon={marker.icon}
+							/>
+						))}
+					</GoogleMap>
+				</LoadScript>
+			)}
+		</Box>
+	);
+};
+
+// Make name optional, and override onChange to not be of `any` type.
+export interface MapProps<T> extends DefaultProps {
+	apiKey: string;
+	data: T[];
+	dataMap: {
+		lat: Array<keyof T>;
```

Comment:
We want these to be an array of keys for somewhat selfish reasons. In balenaCloud we have two fields where we store longitude and latitude, one that is autocalculated, and the other is set by the user. I think it still is a generic solution where you can provide callbacks if a field is not defined (like for title we can have something like [`name`, `slug`].

Reply:
Would an alternative to this array approach be to allow the `dataMap` fields to be either a `keyof T` value or a func that returns a `keyof T` value? That way you can provide a simple, static data map _or_ you can provide callbacks to determine a lat/lng/id/title value.

Reply:
@grahammcculloch yeah makes perfect sense, I was initially thinking that it would be nice if the props are serializable, but I think having it as a function is better 👍